Fix functionality issues

kushal-dreamspring opened the pull request:
- Use Environment Variable for Fixer API Key
- Add alert dialog for errors
- Add config.time_zone to application.rb
- Add date_time_utilities.rb concern
- Use date_time utilities in appointments_controller.rb and doctors_controller.rb
- Fix No Appointments message for signed in user
- Add Locale for hardcoded strings
- Add email validation in _user_form.html.erb
- Remove JSON outputs from controllers
- Add check for Appointment time on cancel appointment action

kushal-dreamspring committed:
Fix invoice_mailer spec

jinifrancis commented:
The 'Next Available' date and time mentioned on the first page are incorrect.

https://github.com/kushal-dreamspring/MediApp/assets/17831956/85406e1e-579b-4285-8aac-fbee1feb2bed

jinifrancis commented:
Missing functionality - I don't see the amount mentioned on the `Pay Now` Button as shown in the Figma design. 
<img width="347" alt="Screenshot 2023-08-04 at 12 47 32 PM" src="https://github.com/kushal-dreamspring/MediApp/assets/17831956/75d3e9ea-a044-4922-85c5-bdf36ba480a2">

<img width="162" alt="Screenshot 2023-08-04 at 12 48 35 PM" src="https://github.com/kushal-dreamspring/MediApp/assets/17831956/6599824c-f02c-49c2-bd20-9caa448225dd">

kushal-dreamspring commented:
> Missing functionality - I don't see the amount mentioned on the `Pay Now` Button as shown in the Figma design. <img alt="Screenshot 2023-08-04 at 12 47 32 PM" width="347" src="https://user-images.githubusercontent.com/17831956/258345612-75d3e9ea-a044-4922-85c5-bdf36ba480a2.png">
> 
> <img alt="Screenshot 2023-08-04 at 12 48 35 PM" width="162" src="https://user-images.githubusercontent.com/17831956/258345986-6599824c-f02c-49c2-bd20-9caa448225dd.png">

Earnest told us earlier to skip this feature

jinifrancis commented:
> > Missing functionality - I don't see the amount mentioned on the `Pay Now` Button as shown in the Figma design. <img alt="Screenshot 2023-08-04 at 12 47 32 PM" width="347" src="https://user-images.githubusercontent.com/17831956/258345612-75d3e9ea-a044-4922-85c5-bdf36ba480a2.png">
> > <img alt="Screenshot 2023-08-04 at 12 48 35 PM" width="162" src="https://user-images.githubusercontent.com/17831956/258345986-6599824c-f02c-49c2-bd20-9caa448225dd.png">
> 
> Earnest told us earlier to skip this feature

Thanks for letting me know about this.

jinifrancis commented:
Looks like there are no validations while creating an appointment. I was able to create an appointment for a doctor for slot 2:00pm at 2:30pm which is incorrect. The backend should prevent creation of appointment

<img width="1728" alt="Screenshot 2023-08-04 at 2 37 49 PM" src="https://github.com/kushal-dreamspring/MediApp/assets/17831956/8fa5722d-31e4-4958-95b3-15a213e58efa">
pointments for slot after that time.

jinifrancis commented:
Why is the `Cancel Appointment` button displayed for appointments past their appointment time?
<img width="1728" alt="Screenshot 2023-08-04 at 2 40 31 PM" src="https://github.com/kushal-dreamspring/MediApp/assets/17831956/e612a9e0-dfd1-4bc2-a5b5-1c81951eec6e">

jinifrancis commented:
I am also able to see that the cancel button is shown for appointments within the 30 mins to the appointment time also. These appointments should not be cancellable. I see backend validations are present. 
<img width="1728" alt="Screenshot 2023-08-04 at 2 49 50 PM" src="https://github.com/kushal-dreamspring/MediApp/assets/17831956/23228b5d-d143-4cba-a5ce-e9e67d7c2dd4">

jinifrancis commented:
Where are the changes to fix the CSV file download? There are still errors in the CSV that can be downloaded on this PR
[appointment 4.csv](https://github.com/kushal-dreamspring/MediApp/files/12259238/appointment.4.csv)

jinifrancis commented:
Validation for name still not worked on. Email validations are messing up the Pay button and is not validated properly.

https://github.com/kushal-dreamspring/MediApp/assets/17831956/a038c83f-5800-4b41-b137-266ff26c7ee3

<img width="376" alt="Screenshot 2023-08-04 at 3 05 27 PM" src="https://github.com/kushal-dreamspring/MediApp/assets/17831956/cd05619c-4e3c-4e1f-aae2-b56bf62e3378">

jinifrancis commented:
Make sure main is merged into this Branch - I had initially asked you to update title in Appointments page to "My Appointments" as in the Figma design. I don't see that change in this PR. Or verify if any change made here is incorrect.
<img width="712" alt="Screenshot 2023-08-04 at 3 08 18 PM" src="https://github.com/kushal-dreamspring/MediApp/assets/17831956/3de6bd67-f8f8-4b9a-b304-2156e3070c98">

kushal-dreamspring commented:
> I am also able to see that the cancel button is shown for appointments within the 30 mins to the appointment time also. These appointments should not be cancellable. I see backend validations are present. <img alt="Screenshot 2023-08-04 at 2 49 50 PM" width="1728" src="https://user-images.githubusercontent.com/17831956/258378835-23228b5d-d143-4cba-a5ce-e9e67d7c2dd4.png">

If you click on the cancel button, it will give an error. I added the backend validation for it, but not on the frontend

jinifrancis commented:
> > I am also able to see that the cancel button is shown for appointments within the 30 mins to the appointment time also. These appointments should not be cancellable. I see backend validations are present. <img alt="Screenshot 2023-08-04 at 2 49 50 PM" width="1728" src="https://user-images.githubusercontent.com/17831956/258378835-23228b5d-d143-4cba-a5ce-e9e67d7c2dd4.png">
> 
> If you click on the cancel button, it will give an error. I added the backend validation for it, but not on the frontend

@kushal-dreamspring I have already mentioned I noticed the backend validation. Please re-read the comment and understand what the intent is here. The expectation here is that the cancel button shouldn't been shown for appointments that cannot be cancelled

jinifrancis commented:
We need the ability for a client to enter email via a form that validates the email to view his/her appoinments. Add this functionality. 

> To view the Invoice/Appointment details we do not need a login system, but just a form to authenticate the user’s email address(only).

jinifrancis commented:
The link on the email is not clickable

<img width="686" alt="Screenshot 2023-08-04 at 4 42 53 PM" src="https://github.com/kushal-dreamspring/MediApp/assets/17831956/40cac102-2c3a-446e-a3ed-48bba360846a">

jinifrancis commented:
I am not able to see this link in the email

> Provide a link in the email, which takes them to the web app where they can see the same Invoice/Appointment details. And see an option to download the Invoice.

kushal-dreamspring committed:
Fix next available appointment in Doctor Card

kushal-dreamspring committed:
Validate date_time in Appointment model

kushal-dreamspring committed:
Fix appointment card

- Show Cancel appointment button only if cancellable
- Fix styling for appointment card

kushal-dreamspring committed:
Fix URL in appointment mail

- Add anchor tag to mail link
- Add html render type to appointment card
- Replace pdf link with app appointment link

kushal-dreamspring committed:
Fix CSV file for appointments

kushal-dreamspring committed:
Fix form validation in User form for new appointments

- Add submit action to display loader
- Add required check to email and name validation
- Add pattern validation to name field

kushal-dreamspring committed:
Add login form for my appointments page

- Add login form page
- Add login action to users_controller.rb
- Add login route to routes.rb
- Redirect to login if user not signed in
- Redirect to my appointments after login
- Update english locale

kushal-dreamspring committed:
Fix minor bugs

- Add amount and currency to appointment mails
- Internationalize mail views and login form

kushal-dreamspring committed:
Update english locale file

kushal-dreamspring committed:
Fix mailer bugs and specs

kushal-dreamspring committed:
Add request specs and minor refactors

- Add appointments_spec.rb
- Add next_appointment test in doctors_spec.rb
- Add /login specs in users_spec.rb
- Minor refactors in doctors_controller.rb, appointments_controller.rb and doctors_controller.rb
- Add gem 'rails-controller-testing'

kushal-dreamspring committed:
Merge branch 'main' into fix/functionality

jinifrancis commented:
Add proper text indicating that numbers or special characters are not allowed when entering a name instead of this generic message
<img width="362" alt="Screenshot 2023-08-16 at 12 27 41 PM" src="https://github.com/kushal-dreamspring/MediApp/assets/17831956/5d9f177a-c943-492b-b291-d6bd33e5019f">

jinifrancis commented:
I am not able to continue with the booking after fixing the validation on the name 

https://github.com/kushal-dreamspring/MediApp/assets/17831956/15142d05-18a9-401b-8153-3d736074fdcb

kushal-dreamspring committed:
Fix CSV file for appointments

- Add comma check for appointment dates

kushal-dreamspring committed:
Fix name validation in _user_form.html.erb

- Add custom error message on name field
- Add space to regex in name field

jinifrancis commented:
When the Fake service runs, the loader is shown on the button but the button is not disabled. This allows for multiple clicks on the button creating multiple appointments with the same doctor with the same timeslot. Make sure button is disabled on click of Book Appointment instead of just showing the loader on the button. Also, it looks like there are missing validations which do not check that the same patient-doctor appointment is created at the same time multiple times.


https://github.com/kushal-dreamspring/MediApp/assets/17831956/340a4f3d-b4ef-4cdc-89ab-6aa035b09b6b

jinifrancis commented:
The link to view the appointment from the email opens the show page for the appointment. It is valid for the current user. But changing the appointment id from the URL, I am able to see all active appointments one by one. Verify that only the person with active session can view his/her appointment and not any other appointment. Also, display proper messaging if an appointment is not found.


https://github.com/kushal-dreamspring/MediApp/assets/17831956/68880435-55bb-4e02-b11f-ac2333ced667

This is also an issue, because one person can cancel another person's appointment. Make sure that only the person whose session is active can view/cancel his/her appointment.

kushal-dreamspring committed:
Fix new appointment submit action

- Add uniqueness check on date_time for each user
- Add uniqueness check on date_time for each doctor
- Add loading payment page while fake_service is running
- Remove redundant data-action from pay button

kushal-dreamspring committed:
Merge 'main' into 'fix/functionality'

kushal-dreamspring committed:
Add appointment validation for show and delete actions

- Add authorise user check to show and delete actions
- Add appointment present check for show and delete actions

jinifrancis commented:
Add an error message indicating that user already has an appointment at a particular time slot. Backend validations are already present, but the display of a proper error message is useful in this scenario.


https://github.com/kushal-dreamspring/MediApp/assets/17831956/c5e82376-c751-4993-b231-ebaf7433230d

kushal-dreamspring committed:
Merge branch 'main' into fix/functionality

kushal-dreamspring committed:
Add error message indicating that user already has an appointment

kushal-dreamspring committed:
Refactor currency_helper.rb

- Write dummy data to cache before read
- Use cache in currency_helper_spec.rb
